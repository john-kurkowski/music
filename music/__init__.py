"""Miscellaneous tasks for publishing my music."""
