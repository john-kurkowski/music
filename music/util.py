"""Misc. utilities."""

from collections.abc import Iterator
from typing import NoReturn

import reapy


def assert_exhaustiveness(no_return: NoReturn) -> NoReturn:
    """Provide an assertion at type-check time that this function is never called."""
    raise AssertionError(f"Invalid value: {no_return!r}")


def find_project() -> reapy.core.Project:
    """Find the target Reaper project."""
    try:
        return reapy.Project()
    except AttributeError as aterr:
        if "module" in str(aterr) and "reascript_api" in str(aterr):
            raise Exception(
                "Error while loading Reaper project. Is Reaper running?"
            ) from aterr
        raise


def set_param_value(param: reapy.core.FXParam, value: float) -> None:
    """Set a parameter's value. Work around bug with reapy 0.10's setter."""
    parent_fx = param.parent_list.parent_fx
    parent = parent_fx.parent
    param.functions["SetParamNormalized"](  # type: ignore[operator]
        parent.id, parent_fx.index, param.index, value
    )


def track_and_parents(track: reapy.core.Track) -> Iterator[reapy.core.Track]:
    """List a track and all of its parent tracks."""
    next_track: reapy.Track | None = track
    while next_track:
        yield next_track
        next_track = next_track.parent_track
