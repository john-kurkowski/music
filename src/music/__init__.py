"""Tasks for publishing my music."""
