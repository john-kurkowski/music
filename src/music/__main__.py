#!/usr/bin/env python


"""CLI for this package."""

import importlib
from pathlib import Path

import click


@click.group(name="music")
def cli() -> None:
    """Tasks for publishing my music."""


for command_file in Path(__file__).parent.glob("*/command.py"):
    command_package = command_file.parent.name
    command_module = importlib.import_module(f"music.{command_package}.command")
    cli.add_command(command_module.main)


if __name__ == "__main__":  # pragma: no cover
    cli()
