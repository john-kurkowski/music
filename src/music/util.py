"""Misc. utilities."""

import asyncio
import enum
import json
import os
import shutil
import warnings
from collections.abc import Callable, Iterator
from functools import wraps
from pathlib import Path
from typing import Any, NoReturn, TypeVar, cast

import aiohttp

with warnings.catch_warnings():
    warnings.filterwarnings("ignore", message="Can't reach distant API")
    import reapy

T = TypeVar("T")

# File: Render project, using the most recent render settings, auto-close render dialog
RENDER_CMD_ID = 42230


class SongVersion(enum.Enum):
    """Different versions of a song to render."""

    MAIN = enum.auto()
    INSTRUMENTAL = enum.auto()
    ACAPPELLA = enum.auto()
    STEMS = enum.auto()

    def name_for_project_dir(self, project_dir: Path) -> str:
        """Name of the project for the given song version."""
        project_name = project_dir.name
        if self is SongVersion.MAIN:
            return project_name
        elif self is SongVersion.INSTRUMENTAL:
            return f"{project_name} (Instrumental)"
        elif self is SongVersion.ACAPPELLA:
            return f"{project_name} (A Cappella)"
        elif self is SongVersion.STEMS:
            return f"{project_name} (Stems)"
        else:  # pragma: no cover
            assert_exhaustiveness(self)

    def path_for_project_dir(self, project_dir: Path) -> Path:
        """Path of the rendered file for the given song version."""
        basename = project_dir / self.name_for_project_dir(project_dir)
        if self is SongVersion.STEMS:
            return basename

        return basename.with_suffix(".wav")

    @property
    def pattern(self) -> list[Path]:
        """Reaper directory render pattern for the given song version, if any."""
        if self is SongVersion.STEMS:
            # Roughly create a directory tree matching the tracks and folders in the Reaper project.
            return [Path("$folders $tracknumber - $track")]

        return []


class ExtendedProject(reapy.core.Project):
    """Extend reapy.core.Project with additional properties."""

    def __init__(self) -> None:
        """Wrap common error in a more helpful message."""
        try:
            super().__init__()
        except AttributeError as aterr:
            if "module" in str(aterr) and "reascript_api" in str(aterr):
                raise Exception(
                    "Error while loading Reaper project. Is Reaper running?"
                ) from aterr
            raise  # pragma: no cover

    @classmethod
    def get_or_open(cls, project_dir: Path) -> "ExtendedProject":
        """Open the target Reaper project if it is not already open."""
        project = cls()
        if project_dir is None or str(project_dir.resolve()) == project.path:
            return project

        project_file = (
            project_dir
            if project_dir.suffix == ".rpp"
            else project_dir / f"{project_dir.name}.rpp"
        )
        reapy.RPR.Main_openProject(str(project_file))  # type: ignore[attr-defined]
        return cls()

    @property
    def metadata(self) -> dict[str, Any]:
        """Parse optional author-idiosyncratic metadata from the project notes."""
        notes = reapy.RPR.GetSetProjectNotes(-1, False, "", 999)[2]  # type: ignore[attr-defined]
        try:
            di = json.loads(notes)
        except json.JSONDecodeError:
            return {}

        return cast(dict[str, Any], di)

    async def render(self) -> None:
        """Trigger Reaper to render the currently open project.

        Unlike sending a command via Reaper's Python API
        (`project.perform_action(action_id)`), this method uses Reaper's HTTP
        API, to work async.
        """
        port = reapy.config.WEB_INTERFACE_PORT

        async with aiohttp.ClientSession() as client:
            resp = await client.get(
                f"http://localhost:{port}/_/{RENDER_CMD_ID}",
                timeout=aiohttp.ClientTimeout(total=60 * 30),
            )
            resp.raise_for_status()

    @property
    def path(self) -> str:
        """Override. Get the path containing the project.

        Works around a bug in reapy 0.10.0's implementation of `Project.path`,
        which actually gets the _recording_ path of the project.
        """
        filename = str(reapy.RPR.EnumProjects(-1, None, 999)[2])  # type: ignore[attr-defined]
        return str(Path(filename).parent)


def assert_exhaustiveness(no_return: NoReturn) -> NoReturn:  # pragma: no cover
    """Provide an assertion at type-check time that this function is never called."""
    raise AssertionError(f"Invalid value: {no_return!r}")


def coro(f: Callable[..., Any]) -> Callable[..., Any]:
    """Decorate Click commands as coroutines.

    H/T https://github.com/pallets/click/issues/85
    """

    @wraps(f)
    def wrapper(*args: Any, **kwargs: Any) -> Any:
        return asyncio.run(f(*args, **kwargs))

    return wrapper


def recurse_property(prop: str, obj: T | None) -> Iterator[T]:
    """Recursively yield the given optional, recursive property, starting with the given object."""
    while obj is not None:
        yield obj
        obj = getattr(obj, prop, None)


def rm_rf(path: Path) -> None:
    """Delete a file or directory recursively, if it exists, similarly to `rm -rf <PATH>`."""
    if os.path.isdir(path) and not os.path.islink(path):
        shutil.rmtree(path)
    elif os.path.exists(path):
        os.remove(path)


def set_param_value(param: reapy.core.FXParam, value: float) -> None:
    """Set a parameter's value.

    Works around bug with reapy 0.10's setter.
    """
    parent_fx = param.parent_list.parent_fx
    parent = parent_fx.parent
    param.functions["SetParamNormalized"](  # type: ignore[operator]
        parent.id, parent_fx.index, param.index, value
    )
