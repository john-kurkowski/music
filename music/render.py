"""Render vocal and instrumental versions of the current Reaper project."""

from collections.abc import Container
import enum
import pathlib
import random
import shutil
import subprocess

import click
import reapy

from .util import assert_exhaustiveness, find_project, set_param_value

LIMITER_RANGE = sum(abs(point) for point in (-60.0, 12.0))

VOCAL_LOUDNESS_WORTH = 2.0

# File: Render project, using the most recent render settings, auto-close render dialog
RENDER_CMD_ID = 42230


class SongVersion(enum.Enum):
    """Different versions of a song to render."""

    MAIN = enum.auto()
    INSTRUMENTAL = enum.auto()


def find_master_limiter_threshold(project: reapy.core.Project) -> reapy.core.FXParam:
    """Find the master track's master limiter's threshold parameter."""
    limiters = [fx for fx in project.master_track.fxs[::-1] if "Limit" in fx.name]
    if not limiters:
        raise ValueError("Master limiter not found")
    limiter = limiters[0]

    def safe_param_name(param: reapy.core.FXParam) -> str:
        """Work around uncaught exception for non-UTF-8 strings in parameter names."""
        try:
            return param.name
        except reapy.errors.DistError as ex:
            if "UnicodeDecodeError" in str(ex):
                return ""
            raise

    thresholds = [
        param for param in limiter.params if "Threshold" in safe_param_name(param)
    ]
    return thresholds[0]


def print_summary_stats(fil: pathlib.Path) -> None:
    """Print statistics for the given audio file, like LUFS-I and LRA."""
    cmd: list[str | pathlib.Path] = [
        "ffmpeg",
        "-i",
        fil,
        "-filter:a",
        ",".join(("volumedetect", "ebur128=framelog=verbose")),
        "-f",
        "null",
        "/dev/null",
    ]
    with subprocess.Popen(cmd):
        pass


def render_version(project: reapy.core.Project, version: SongVersion) -> None:
    """Trigger Reaper to render the current project audio. Name the output file
    according to the given version."""
    project_name = pathlib.Path(project.name).stem
    if version is SongVersion.MAIN:
        out_name = project_name
    elif version is SongVersion.INSTRUMENTAL:
        out_name = f"{project_name} (Instrumental)"
    else:
        assert_exhaustiveness(version)

    # Avoid "Overwrite" "Render Warning" dialog, which can't be scripted, with a temporary filename
    rand_id = random.randrange(10**5, 10**6)
    in_name = f"{out_name} {rand_id}.tmp"

    project.set_info_string("RENDER_PATTERN", in_name)
    try:
        project.perform_action(RENDER_CMD_ID)
    finally:
        project.set_info_string("RENDER_PATTERN", "$project")

    out_dir = pathlib.Path(project.path)
    out_fil = out_dir / f"{out_name}.wav"
    shutil.move(out_dir / f"{in_name}.wav", out_fil)

    print_summary_stats(out_fil)


def main(versions: Container[SongVersion] | None = None) -> None:
    """Module entrypoint."""

    if versions is None:
        versions = set(SongVersion)

    project = find_project()

    threshold = find_master_limiter_threshold(project)
    threshold_previous_value = threshold.normalized
    threshold_louder_value = (
        (threshold_previous_value * LIMITER_RANGE) - VOCAL_LOUDNESS_WORTH
    ) / LIMITER_RANGE

    vocals = next((track for track in project.tracks if track.name == "Vocals"), None)

    did_something = False

    if SongVersion.MAIN in versions:
        did_something = True
        if vocals:
            vocals.unmute()
        render_version(project, SongVersion.MAIN)

    if SongVersion.INSTRUMENTAL in versions and vocals:
        did_something = True
        try:
            set_param_value(threshold, threshold_louder_value)
            vocals.mute()
            render_version(project, SongVersion.INSTRUMENTAL)
        finally:
            set_param_value(threshold, threshold_previous_value)
            vocals.unmute()

    if not did_something:
        raise click.UsageError("nothing to render")

    # Render causes a project to have unsaved changes, no matter what. Save the user a step.
    project.save()


if __name__ == "__main__":
    main()
