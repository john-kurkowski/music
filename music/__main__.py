#!/usr/bin/env python


"""CLI for this package."""

import click

from .render import SongVersion, main as _render


@click.group()
def cli() -> None:
    """Miscellaneous tasks for publishing my music."""


@cli.command()
@click.option(
    "--include-main",
    default=None,
    flag_value=SongVersion.MAIN,
    help=(
        "Whether to render the main version. Defaults to rendering all versions, unless"
        ' one of the "--include-*" flags is set.'
    ),
    type=SongVersion,
)
@click.option(
    "--include-instrumental",
    default=None,
    flag_value=SongVersion.INSTRUMENTAL,
    help=(
        "Whether to render the instrumental version. Rendering this version is skipped"
        " if no vocals exist. Defaults to rendering all versions, unless one of the"
        ' "--include-*" flags is set.'
    ),
    type=SongVersion,
)
def render(
    include_main: SongVersion | None, include_instrumental: SongVersion | None
) -> None:
    """Render vocal, instrumental versions of the current Reaper project.
    Overwrites existing versions. Note the Reaper preference "Set media items
    offline when application is not active" should be unchecked, or media items
    will be silent in the render."""
    versions = {
        version for version in (include_main, include_instrumental) if version
    } or None
    _render(versions)


if __name__ == "__main__":
    cli()
