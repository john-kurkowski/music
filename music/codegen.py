"""Generate code for this package."""

import os
import subprocess
from pathlib import Path

import openai

from .render import _cmd_for_stats


def main(example_audio_file: Path) -> None:
    """Generate a parser for the stats output of ffmpeg.

    ffmpeg's output is verbose and not easily parsed. A parser would be brittle
    to maintain by hand. Have AI write it for us.
    """
    openai.api_key = os.environ["OPENAI_API_KEY"]

    fn_name = "parse_summary_stats"
    cmd = _cmd_for_stats(example_audio_file)
    proc = subprocess.run(cmd, check=True, stderr=subprocess.PIPE, text=True)
    proc_output = proc.stderr

    messages = [
        {
            "role": "system",
            "content": (
                "You are a CLI that writes Python source code. Do not converse. Do not"
                " show examples. Only respond in Python source code. Always include"
                " function argument and return type annotations. Disable type checking"
                " on lines that return Optionals."
            ),
        },
        {
            "role": "user",
            "content": (
                f"Write a Python function named {fn_name} to parse the following output"
                " for max volume, LUFS-I, and LRA. Use regular expressions. Return the"
                f" values in a dict.\n\n{proc_output}"
            ),
        },
    ]
    response = openai.ChatCompletion.create(model="gpt-3.5-turbo", messages=messages, temperature=0)  # type: ignore[no-untyped-call]
    lines = response.choices[0].message["content"].splitlines()

    if lines[0].startswith("```"):
        lines = lines[1:]
    if lines[-1].startswith("```"):
        lines = lines[:-1]

    out_fil = Path(__file__).parent / "__codegen__" / "stats.py"
    out_fil.write_text("\n".join(lines))

    print(response.usage)
