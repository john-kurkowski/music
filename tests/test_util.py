"""Util tests."""

from pathlib import Path
from unittest import mock

import music.util
import pytest
from syrupy.assertion import SnapshotAssertion


@mock.patch("reapy.core.Project.__init__")
def test_project_reaper_not_running(core_project: mock.Mock) -> None:
    """Test handling when Reaper is not running."""
    with pytest.raises(Exception, match="Reaper running"):
        core_project.side_effect = AttributeError(
            "module doesn't have reascript_api, yo"
        )
        music.util.ExtendedProject()


@mock.patch("reapy.core.Project.__init__")
@mock.patch("reapy.RPR.EnumProjects", create=True)
def test_project_path(
    enum_projects: mock.Mock,
    core_project: mock.Mock,
    tmp_path: Path,
) -> None:
    """Test getting the path containing the project (not the _recording_ path)."""
    some_path = tmp_path / "path" / "to" / "some project"
    some_file = some_path / "some project.rpp"
    enum_projects.return_value = ("", "", some_path / some_file)
    obj = music.util.ExtendedProject()
    assert obj.path == str(some_path)


@mock.patch("reapy.core.Project.__init__")
@mock.patch("reapy.RPR.EnumProjects", create=True)
@mock.patch("reapy.RPR.Main_openProject", create=True)
def test_project_open(
    open_project: mock.Mock,
    enum_projects: mock.Mock,
    core_project: mock.Mock,
    snapshot: SnapshotAssertion,
    tmp_path: Path,
) -> None:
    """Test invoking Reaper to open a project."""
    some_path = tmp_path / "path" / "to" / "some project"
    music.util.ExtendedProject(some_path)
    assert open_project.mock_calls == snapshot
