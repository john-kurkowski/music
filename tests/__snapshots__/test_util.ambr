# serializer version: 1
# name: test_project_open
  _CallList([
    _Call(
      '',
      tuple(
        'TMP_PATH_HERE/path/to/some project/some project.rpp',
      ),
      dict({
      }),
    ),
  ])
# ---
# name: test_project_render
  dict({
    'get': _CallList([
      _Call(
        '',
        tuple(
          'http://localhost:2307/_/42230',
        ),
        dict({
        }),
      ),
    ]),
    'post': _CallList([
    ]),
    'put': _CallList([
    ]),
  })
# ---
